listener.go

```diff
@@ -0,0 +1,112 @@
+package connmux
+
+import (
+	"io"
+	"net"
+)
+
+type listener struct {
+	wrapped net.Listener
+	pool    BufferPool
+	errCh   chan error
+	connCh  chan net.Conn
+}
+
+// WrapListener wraps the given listener with support for multiplexing. Only
+// connections that start with the special session start sequence will be
+// multiplexed, otherwise connections behave as normal. This means that a single
+// listener can be used to serve clients that do multiplexing as well as other
+// clients that don't.
+//
+// Multiplexed sessions can only be initiated immediately after opening a
+// connection to the Listener.
+//
+// pool - BufferPool to use
+func WrapListener(wrapped net.Listener, pool BufferPool) net.Listener {
+	l := &listener{
+		wrapped: wrapped,
+		pool:    pool,
+		connCh:  make(chan net.Conn),
+		errCh:   make(chan error),
+	}
+	go l.process()
+	return l
+}
+
+func (l *listener) Accept() (net.Conn, error) {
+	select {
+	case conn := <-l.connCh:
+		return conn, nil
+	case err := <-l.errCh:
+		return nil, err
+	}
+}
+
+func (l *listener) Addr() net.Addr {
+	return l.wrapped.Addr()
+}
+
+func (l *listener) Close() error {
+	go func() {
+		l.errCh <- ErrListenerClosed
+	}()
+	// Closing wrapped has the side effect of making the process loop terminate
+	// because it will fail to accept from wrapped.
+	return l.wrapped.Close()
+}
+
+func (l *listener) process() {
+	for {
+		conn, err := l.wrapped.Accept()
+		if err != nil {
+			l.errCh <- err
+			return
+		}
+		go l.onConn(conn)
+	}
+}
+
+func (l *listener) onConn(conn net.Conn) {
+	b := make([]byte, sessionStartTotalLen)
+	// Try to read start sequence
+	_, err := io.ReadFull(conn, b)
+	if err != nil {
+		l.errCh <- err
+		return
+	}
+	if string(b[:sessionStartHeaderLen]) == sessionStart {
```

Comment:
It's possible that a normal connection starts with the sequence. Unlikely to happen, but maybe worth to mention in the package doc